# Linear Programming using CVXPY

In [1]:
# Make sure you have the package cvxpy installed
import cvxpy as cp
import numpy as np

# Exercise 1

A carpenter makes tables and chairs. Each table can be sold for a profit of £30 and each chair for a profit of £10. The carpenter can afford to spend up to 40 hours per week working and takes six hours to make a table and three hours to make a chair. Customer demand requires that he makes at least three times as many chairs as tables. Tables take up four times as much storage space as chairs and there is room for at most four tables each week. 

## Solution

Let x be the number of chairs and y the number of tables. Let us determine the x and y that maximize the profit.
We want to maximize 10x+30y under the conditions 3x+6y $\le$ 40; x $\ge$ 3y;        
4x+y $\le$ 16; x$\ge$ 0; y$\ge$0

In [9]:
x = cp.Variable()
y=cp.Variable()
c=10*x+30*y
prob = cp.Problem(cp.Maximize(c),
                 [3*x+6*y<=40, x>=3*y, x+4*y<=16])
prob.solve()

# Print result.
print("\nThe optimal value is", prob.value)
print("A solution is")
print("x=", x.value)
print("y=", y.value)


The optimal value is 146.66666666666669
A solution is
x= 10.666666666666663
y= 1.333333333333335


The maximum profit they can make is 146.67£ and corresponds to arround 11 chairs and 1 table.

# Exercise 2

## Solution

Let x be the number of products of type A and y the number of products of type B. Let us determine the x and y that maximize the profit.
We want to maximize 3x+5y under the conditions 12x+25y $\le$ 1800; $\frac{y}{2}$ $\ge$ $\frac{x}{5}$; x$\ge$0; y$\ge$0

In [16]:
x = cp.Variable()
y=cp.Variable()
c=3*x+5*y
prob = cp.Problem(cp.Maximize(c),
                 [12*x+25*y<=1800, 5*y>=2*x])
prob.solve()

# Print result.
print("\nThe optimal value is", prob.value)
print("A solution is")
print("x=", x.value)
print("y=", y.value)


The optimal value is 409.0909090909091
A solution is
x= 81.8181818181818
y= 32.72727272727273


The company hires a new machine

In [17]:
x = cp.Variable()
y=cp.Variable()
c=3*x+5*y
prob = cp.Problem(cp.Maximize(c),
                 [12*x+25*y<=3600, 5*y>=2*x])
prob.solve()

# Print result.
print("\nThe new optimal value is", prob.value)
print("A solution is")
print("x=", x.value)
print("y=", y.value)


The new optimal value is 818.1818181818182
A solution is
x= 163.6363636363636
y= 65.45454545454545


They will at most afford to pay an amount less than 818.18-409.09=409.09$.

# Exercise 3

## Solution

Let x be the number of videos sold, and y the number of printed books sold. Let us determine the x and y that maximize the profit.
We want to maximize 50x+30y under the conditions x+y $\le 10000$; $x$ $\ge$ $4000$; y$\ge$  2000; y$\le$4000

In [18]:
x = cp.Variable()
y=cp.Variable()
c=50*x+30*y
prob = cp.Problem(cp.Maximize(c),
                 [x+y<=10000, x>=4000, y>=2000, y<=4000])
prob.solve()

# Print result.
print("\nThe optimal value is", prob.value)
print("A solution is")
print("x=", x.value)
print("y=", y.value)


The optimal value is 460000.0
A solution is
x= 8000.0
y= 2000.0


The maximum income is 460000£ and corresponds to 8000 videos and 2000 books sold.